realm/realm-library/src/objectServer/java/io/realm/SyncSession.java

```diff
@@ -71,29 +90,26 @@ public SyncUser getUser() {
      * @return {@link URI} describing the remote Realm.
      */
     public URI getServerUrl() {
-        return osSession.getConfiguration().getServerUrl();
+        return configuration.getServerUrl();
     }
 
-    /**
-     * Returns the state of this session.
-     *
-     * @return the current {@link SessionState} for this session.
-     */
-    public SessionState getState() {
-        return osSession.getState();
-    }
-
-    ObjectServerSession getOsSession() {
-        return osSession;
+    // Called from native code.
+    // This callback will happen on the thread running the Sync Client.
+    @KeepMember
+    void notifySessionError(int errorCode, String errorMessage) {
+        ObjectServerError error = new ObjectServerError(ErrorCode.fromInt(errorCode), errorMessage);
+        if (errorHandler != null) {
+            errorHandler.onError(this, error);
+        }
     }
 
-    @Override
-    protected void finalize() throws Throwable {
-        super.finalize();
-        if (osSession.getState() != SessionState.STOPPED) {
-            RealmLog.warn("Session was not closed before being finalized. This is a potential resource leak.");
-            osSession.stop();
+    // This is called from a synchronized block
```

Comment:
I have a hunch we need to make the lifecyle methods of a Session available when we expose a `SyncPolicy` so if we are forced to do it now it wouldn't be that harmful, although less ideal.